File: include/solver_code.cpp
#include <iostream>
#include <Eigen/Dense>
#include <vector>

using namespace Eigen;
using namespace std;

// Matrix functions
Matrix<double, 3, 4> gMatrix(const Vector4d &p) {
    Matrix<double, 3, 4> G;
    G << -p(1), p(0), -p(3), p(2),
         -p(2), p(3), p(0), -p(1),
         -p(3), -p(2), p(1), p(0);
    return G;
}

Matrix<double, 3, 4> lMatrix(const Vector4d &p) {
    Matrix<double, 3, 4> L;
    L << -p(1), p(0), p(3), -p(2),
         -p(2), -p(3), p(0), p(1),
         -p(3), p(2), -p(1), p(0);
    return L;
}

Matrix4d skewN(const Vector3d &a) {
    Matrix4d Sn = Matrix4d::Zero();
    Sn.block<1, 3>(0, 1) = -a.transpose();
    Sn.block<3, 1>(1, 0) = a;
    Sn.block<3, 3>(1, 1) = -Matrix3d((Matrix3d() << 0, -a(2), a(1),
                                                    a(2), 0, -a(0),
                                                   -a(1), a(0), 0).finished());
    return Sn;
}

int main() {
    // Constants
    constexpr double m1 = 10.0, l1 = 2.0, h1 = 0.7, w1 = 0.7;

    constexpr double Ix = (1.0 / 12.0) * m1 * (h1 * h1 + w1 * w1);
    constexpr double Iy = (1.0 / 12.0) * m1 * (l1 * l1 + w1 * w1);
    constexpr double Iz = (1.0 / 12.0) * m1 * (l1 * l1 + h1 * h1);

    constexpr double totalTime = 10.0, timeStep = 0.001;

    int steps = static_cast<int>(totalTime / timeStep);

    // Initial Conditions
    VectorXd q1(7), q1d(7);
    q1 << 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0;
    q1d.setZero();

    // Vector3d s0A = Vector3d::Zero();
    Vector3d s1Am(-l1/2, 0.0, 0.0);  // s1Bm(L2, 0.0, 0.0);

    // Mass and Inertia matrix
    Matrix3d N1 = (Matrix3d() << m1, 0, 0, 0, m1, 0, 0, 0, m1).finished();
    Matrix3d J1m = (Matrix3d() << Ix, 0, 0, 0, Iy, 0, 0, 0, Iz).finished();

    // Storage
    vector<Vector3d> positions(steps);
    vector<Vector4d> orientations(steps);

    for (int step = 0; step < steps; ++step) {
        constexpr double g = 9.81;
        Vector3d r1 = q1.head<3>();
        Vector4d p1 = q1.segment<4>(3);
        // Vector3d r1d = q1d.head<3>();
        Vector4d p1d = q1d.segment<4>(3);

        // Inertia and Constraint Matrices
        Matrix<double, 3, 4> L1 = lMatrix(p1);
        Matrix4d J1s = 4 * L1.transpose() * J1m * L1;
        MatrixXd Ms(7, 7);
        Ms.setZero();
        Ms.topLeftCorner<3, 3>() = N1;
        Ms.bottomRightCorner<4, 4>() = J1s;

        Matrix<double, 3, 7> B;
        B.setZero();
        B.topLeftCorner<3, 3>() = -Matrix3d::Identity();
        B.rightCols<4>() = -2 * gMatrix(p1) * skewN(s1Am);

        Matrix<double, 3, 4> L1d = lMatrix(p1d);
        Vector3d gamma = -(-2 * gMatrix(p1d) * L1d.transpose() * s1Am);

        Matrix<double, 1, 7> P;
        P.setZero();
        P.rightCols<4>() = p1.transpose();

        MatrixXd A(11, 11);
        A.setZero();
        A.topLeftCorner<7, 7>() = Ms;
        A.block<1, 7>(7, 0) = P;
        A.block<7, 1>(0, 7) = P.transpose();
        A.block<3, 7>(8, 0) = B;
        A.block<7, 3>(0, 8) = B.transpose();

        VectorXd bs(7);
        bs.setZero();
        bs.tail<4>() = 2 * (4 * L1d.transpose() * J1m * L1 * p1d);

        double c = p1d.squaredNorm();
        VectorXd Bv(11);
        Bv.setZero();
        Bv.segment<7>(0) = bs;
        Bv(7) = c;

        Vector3d f1(0.0, 0.0, m1 * g);
        Vector4d n1s = Vector4d::Zero();

        VectorXd gs(11);
        gs << f1, n1s, 0, gamma;

        VectorXd x = A.partialPivLu().solve(gs - Bv);
        VectorXd q1dd = x.segment<7>(0);

        // Integration
        q1d += q1dd * timeStep;
        q1 += q1d * timeStep;

        // Normalize quaternion
        p1 = q1.segment<4>(3);
        q1.segment<4>(3) = p1 / p1.norm();

        // Store results
        positions[step] = r1;
        orientations[step] = q1.segment<4>(3);

    }

    // Print Final Results
    for (int i = 0; i < steps; i += steps / 100) {
        cout << "Time Step: " << i * timeStep << endl;
        cout << "Position: " << positions[i].transpose() << endl;
        cout << "Orientation: " << orientations[i].transpose() << endl;
        cout << "--------------------------" << endl;
    }

    return 0;
}

File: include/solver_code.m
clear; close all; clc; format shortG;

% 11.42
% Table 7.1

% Constaints
L1 = 1.0;
L2 = 1.0;

s0A = [0; 0; 0];
s1Am = [-L1; 0; 0];
s1Bm = [L2; 0; 0];


% Constants
g = 9.81;               % Gravity
dt = 0.01;              % Time step
t_total = 10;           % Total simulation time
steps = t_total / dt;   % Number of time steps

% Initial Conditions
q1 = [1; 0; 0; 1; 0; 0; 0]; % Initial position and orientation
q1d = [0; 0; 0; 0; 0; 0; 0]; % Initial velocity
r1 = q1(1:3);
p1 = q1(4:7);


% J = [-eye(3), -2 * (gMatrix(p1) * skewN(s1Am) + s1Am * p1')]
% Jacobian = [-eye(3), - 2 * gMatrix(p1) * skewN(s1Am)]

% Gamma = -(-2 * gMatrix(p1d) * lMatrix(p1d)' * s1Am)

% Parameters
N1 = diag([10, 10, 10]);
J1m = diag([0.1, 0.1, 3]);

% Store results for visualization
time = (0:dt:t_total)';
positions = zeros(length(time), 3);
orientations = zeros(length(time), 4);

for step = 1:length(time)
    % Extract position and orientation components
    r1 = q1(1:3);
    p1 = q1(4:7);
    r1d = q1d(1:3);
    p1d = q1d(4:7);
    
    % L1 Matrix
    L1 = [-p1(2),  p1(1),  p1(4), -p1(3);
          -p1(3), -p1(4),  p1(1),  p1(2);
          -p1(4),  p1(3), -p1(2),  p1(1)];
    
    % Generalized Inertia Matrix
    J1s = 4 * L1' * J1m * L1;
    Ms = blkdiag(N1, J1s);
    
    % Jacobian & Gamma
    B = [-eye(3), - 2 * gMatrix(p1) * skewN(s1Am)];

    gamma = -(-2 * gMatrix(p1d) * lMatrix(p1d)' * s1Am);

    % Constraint Matrix
    P = [zeros(1, 3), p1'];
    A = [Ms, P', B'; P, zeros(1,4); B, zeros(3, 4)];
    
    % Time Derivative of L1
    L1d = [-p1d(2),  p1d(1),  p1d(4), -p1d(3);
           -p1d(3), -p1d(4),  p1d(1),  p1d(2);
           -p1d(4),  p1d(3), -p1d(2),  p1d(1)];
    
    % Skew-Symmetric H1
    H1 = 4 * L1d' * J1m * L1;
    bs = [zeros(3, 1); 2 * H1 * p1d];
    
    % Additional Constraint
    c = p1d' * p1d;
    B = [bs; c; 0; 0; 0];
    
    % External Forces and Torques
    f1 = [0; 0; 10];
    n1s = [zeros(4, 1)];
    gs = [f1; n1s];
    c = [gs; 0; gamma];
    
    % Solve for accelerations and constraint force
    x = A \ (c - B);
    q1dd = x(1:7);  % Accelerations
    omega = x(8);   % Constraint force (not used in updates)
    
    % Update velocities and positions using Euler integration
    q1d = q1d + q1dd * dt; % Update velocity
    q1 = q1 + q1d * dt;    % Update position
    
    % Normalize the quaternion (p1 is the quaternion part of q1)
    p1 = q1(4:7);          % Extract quaternion
    q1(4:7) = p1 / norm(p1); % Normalize quaternion
    
    % Store results for visualization
    positions(step, :) = r1';
    orientations(step, :) = q1(4:7)';
end

% Visualization
figure;
subplot(2, 1, 1);
plot(time, positions);
xlabel('Time (s)');
ylabel('Position (m)');
legend('x', 'y', 'z');
title('Position over Time');

subplot(2, 1, 2);
plot(time, orientations);
xlabel('Time (s)');
ylabel('Orientation Components');
legend('p1', 'p2', 'p3', 'p4');
title('Orientation over Time');

function G = gMatrix(p)
    G = [-p(2),  p(1), -p(4),  p(3);
         -p(3),  p(4),  p(1), -p(2);
         -p(4), -p(3),  p(2),  p(1)];
end

function L = lMatrix(p)
    L = [-p(2),  p(1),  p(4), -p(3);
         -p(3), -p(4),  p(1),  p(2);
         -p(4),  p(3), -p(2),  p(1)];
end

function Sn = skewN(a)
    Sn = [0, -a'; a, -skew(a)];
end

function S = skew(v)
    S = [    0, -v(3),  v(2); 
          v(3),     0, -v(1); 
         -v(2),  v(1),     0];
end
